web3/middleware/async_middleware/async_cache.py

```diff
@@ -0,0 +1,104 @@
+import threading
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Collection,
+    Dict,
+    Tuple,
+    Type,
+    cast,
+)
+
+import lru
+
+from web3._utils.caching import (
+    generate_cache_key,
+)
+from web3.types import (
+    AsyncMiddleware,
+    Middleware,
+    RPCEndpoint,
+    RPCResponse,
+)
+
+if TYPE_CHECKING:
+    from web3 import Web3  # noqa: F401
+
+
+SIMPLE_CACHE_RPC_WHITELIST = cast(
+    Tuple[RPCEndpoint],
+    (
+        "web3_clientVersion",
+        "eth_getBlockTransactionCountByHash",
+        "eth_getUncleCountByBlockHash",
+        "eth_getBlockByHash",
+        "eth_getTransactionByHash",
+        "eth_getTransactionByBlockHashAndIndex",
+        "eth_getRawTransactionByHash",
+        "eth_getUncleByBlockHashAndIndex",
+        "eth_chainId",
+    ),
+)
+
+
+def _should_cache_response(
+    _method: RPCEndpoint, _params: Any, response: RPCResponse
+) -> bool:
+    return (
+        "error" not in response
+        and "result" in response
+        and response["result"] is not None
+    )
+
+
+async def async_construct_simple_cache_middleware(
+    cache_class: Type[Dict[Any, Any]],
+    rpc_whitelist: Collection[RPCEndpoint] = SIMPLE_CACHE_RPC_WHITELIST,
+    should_cache_fn: Callable[
+        [RPCEndpoint, Any, RPCResponse], bool
+    ] = _should_cache_response,
+) -> Middleware:
+    """
+    Constructs a middleware which caches responses based on the request
+    ``method`` and ``params``
+
+    :param cache_class: Any dictionary-like object
+    :param rpc_whitelist: A set of RPC methods which may have their responses cached.
+    :param should_cache_fn: A callable which accepts ``method`` ``params`` and
+        ``response`` and returns a boolean as to whether the response should be
+        cached.
+    """
+
+    async def async_simple_cache_middleware(
+        make_request: Callable[[RPCEndpoint, Any], Any], async_w3: "Web3"
+    ) -> AsyncMiddleware:
+        cache = cache_class()
+        lock = threading.Lock()
+
+        async def middleware(method: RPCEndpoint, params: Any) -> RPCResponse:
+            if method in rpc_whitelist:
+                with lock:
+                    cache_key = generate_cache_key((method, params))
+                    if cache_key not in cache:
+                        response = await make_request(method, params)
+                        if should_cache_fn(method, params, response):
+                            cache[cache_key] = response
+                        return response
+                    return cache[cache_key]
+            else:
+                return await make_request(method, params)
+
+        return middleware
+
+    return async_simple_cache_middleware
+
+
+async def _async_simple_cache_middleware(
+    make_request: Callable[[RPCEndpoint, Any], Any], async_w3: "Web3"
+):
+    middleware = await async_construct_simple_cache_middleware(
+        cache_class=cast(Type[Dict[Any, Any]], functools.partial(lru.LRU, 256)),
```

Comment:
Interested in feedback on using LRU here vs anything else

Reply:
The problem with LRU is that it's not thread safe, IIRC.

Reply:
We do use it with sync, but I believe the async implementation is currently thread safe. We do have some outstanding work to do with the session cache, though to get it to work just right. See #2446 and the corresponding PR (#2409 I think)

Reply:
The other issue with using LRU with an async method is that it will not await the callback that it calls when it evicts an item. That was really the main reason that I wrote `SessionCache` in request.py for what it is worth. So if you are going to be using a callback with the above it will just return a coroutine (if I remember correctly) instead of executing the async callback. 

If you really want to you could pull SessionCache out of request.py and put it somewhere in _utils then use that, but that would be a decent amount of work. The SessionCache itself is solid, the main issue still outstanding in request.py is how and where we await the close of an evicted async item.